# Laboratorio 1
## Exploratory Data Analysis (EDA)
***
#### Lau Tejada, Yusuri Arciga, Elena Villalobos, Carlos López

In [3]:
# Paquetería necesaria
import sys
import pandas as pd
import numpy as np
import seaborn as sns
import pandas_profiling

In [4]:
sys.version

'3.7.4 (default, Sep 22 2020, 11:44:10) \n[Clang 11.0.3 (clang-1103.0.32.62)]'

In [5]:
# Cargamos datos 
df = pd.read_csv("/Users/ele/Documents/HeavyDataMacair/consumo_agua.csv")

In [6]:
df.head(3)  # Las primeras tres filas de la base de datos

Geo Point  \
0  19.4552601937,-99.1126617526   
1  19.4552601937,-99.1126617526   
2  19.4557195871,-99.1135822797   

                                           Geo Shape  consumo_total_mixto  \
0  {"type": "MultiPolygon", "coordinates": [[[[-9...               159.72   
1  {"type": "MultiPolygon", "coordinates": [[[[-9...                 0.00   
2  {"type": "MultiPolygon", "coordinates": [[[[-9...                 0.00   

   anio             nomgeo  consumo_prom_dom  consumo_total_dom  \
0  2019  Gustavo A. Madero         42.566364             468.23   
1  2019  Gustavo A. Madero         35.936667             107.81   
2  2019  Gustavo A. Madero         24.586000             122.93   

            alcaldia         colonia  consumo_prom_mixto  consumo_total  \
0  GUSTAVO A. MADERO  7 DE NOVIEMBRE               53.24         631.00   
1  GUSTAVO A. MADERO  7 DE NOVIEMBRE                0.00         115.13   
2  GUSTAVO A. MADERO  7 DE NOVIEMBRE                0.00         197.96   

   consumo_prom  consumo_prom_no_dom  bimestre  consumo_total_no_dom    gid  \
0     42.066667                 3.05         3                  3.05  57250   
1     28.782500                 7.32         3                  7.32  57253   
2     32.993333                75.03         3                 75.03  57255   

  indice_des  
0       ALTO  
1      MEDIO  
2    POPULAR

In [7]:
df.describe() # Observar de manera general los datos

consumo_total_mixto     anio  consumo_prom_dom  consumo_total_dom  \
count         62775.000000  71102.0      66282.000000       66282.000000   
mean            174.359929   2019.0         29.132386        1186.263611   
std             312.663596      0.0         64.565925        2771.038307   
min               0.000000   2019.0          0.000000           0.000000   
25%               0.000000   2019.0         18.690547         161.635000   
50%              79.940000   2019.0         26.414248         604.185000   
75%             233.320000   2019.0         36.246563        1261.445000   
max           23404.440000   2019.0       7796.410000       95060.690000   

       consumo_prom_mixto  consumo_total  consumo_prom  consumo_prom_no_dom  \
count        62775.000000   71102.000000  71102.000000         71102.000000   
mean            50.636234    1695.847222    111.217399           126.760172   
std            130.408673    3555.697457   1069.949262          1095.817805   
min              0.000000       0.000000      0.000000             0.000000   
25%              0.000000     340.952500     23.010139             6.275417   
50%             33.451667     896.175000     31.693818            19.280000   
75%             61.216548    1808.902500     45.484917            54.186875   
max          11702.220000  119726.940000  89691.773438         89691.773438   

           bimestre  consumo_total_no_dom           gid  
count  71102.000000          71102.000000  71102.000000  
mean       2.006807            436.060309  35551.500000  
std        0.814393           2126.152162  20525.523757  
min        1.000000              0.000000      1.000000  
25%        1.000000             10.980000  17776.250000  
50%        2.000000             54.055000  35551.500000  
75%        3.000000            230.430000  53326.750000  
max        3.000000         119726.940000  71102.000000

#### 1. ¿Cuántas variables tenemos?

In [8]:
print('Tenemos ' + str(df.shape[1]) + ' variables')

Tenemos 17 variables


#### 2. ¿Cuántas observaciones tenemos?

In [9]:
print('Tenemos ' + str(df.shape[0]) + ' observaciones')

Tenemos 71102 observaciones


#### 3. ¿Cuántas observaciones únicas tenemos por variable?

In [10]:
df.nunique()

Geo Point               22930
Geo Shape               22922
consumo_total_mixto     24339
anio                        1
nomgeo                     17
consumo_prom_dom        52060
consumo_total_dom       47051
alcaldia                   16
colonia                  1340
consumo_prom_mixto      31911
consumo_total           56015
consumo_prom            62214
consumo_prom_no_dom     37440
bimestre                    3
consumo_total_no_dom    27336
gid                     71102
indice_des                  4
dtype: int64

#### 4. ¿Cuántas variables numéricas tenemos?

In [11]:
numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']

In [62]:
newdf = df.select_dtypes(include=numerics)
print('Son ' + str(newdf.shape[1]) + ' variables numericas')

Son 10 variables numericas


In [63]:
df.dtypes

Geo Point                object
Geo Shape                object
consumo_total_mixto     float64
anio                      int64
nomgeo                   object
consumo_prom_dom        float64
consumo_total_dom       float64
alcaldia                 object
colonia                  object
consumo_prom_mixto      float64
consumo_total           float64
consumo_prom            float64
consumo_prom_no_dom     float64
bimestre                 object
consumo_total_no_dom    float64
gid                       int64
indice_des               object
dtype: object

Que son: 
- comumo_total_mixto
- consumo_prom_dom
- consumo_prom_mixto
- consumo_total
- consumo_prom
- consumo_prom_no_dom
- consumo_total_no_dom
- anio
- gid

#### 5. ¿Cuántas variables de fecha tenemos?

In [13]:
date_df = df.select_dtypes(include='datetime')

In [14]:
print('Tenemos ' + str(date_df.shape[1]) + ' variables de tipo fecha')

Tenemos 0 variables de tipo fecha


#### 6. ¿Cuántas variables categóricas tenemos?

In [15]:
df['bimestre'] = df.bimestre.astype(str)

In [16]:
df.dtypes

Geo Point                object
Geo Shape                object
consumo_total_mixto     float64
anio                      int64
nomgeo                   object
consumo_prom_dom        float64
consumo_total_dom       float64
alcaldia                 object
colonia                  object
consumo_prom_mixto      float64
consumo_total           float64
consumo_prom            float64
consumo_prom_no_dom     float64
bimestre                 object
consumo_total_no_dom    float64
gid                       int64
indice_des               object
dtype: object

Las variables que parecen ser categóricas son: 
- nomgeo
- alcaldia
- colonia
- bimestre
- indice_des

#### 7. ¿Cuántas variables de texto tenemos?

No tenemos realmente variables de texto.

#### 8. Data profiling

**Variables numéricas:**

- Tipo de dato: float, integer
- Número de observaciones
- Media, SD
- Cuartiles: 25%, 50% y 75%
- Valor máximo y mínimo
- Número de observaciones únicas
- Top 5 observaciones repetidas
- Número de observaciones con valores faltantes
- ¿Hay redondeos?

In [47]:
def data_profiling(dataframe):
    """
    Crea el data profiling de variables numericas y categoricas
    """
    
    # Primero asignamos un data frame con los nombres de las columnas que saldran
    df_num = pd.DataFrame(columns = ['tipo_de_dato','observaciones','media','cuantil_25',
                                   'cuantil_50','cuantil_75','valor_max','valor_min',
                                   'observaciones_unicas','top_1','top_2','top_3','numero_de_nan'])
    df_cat = pd.DataFrame(columns = ['n_de_categorias','moda','top_1','top_2','top_3','numero_de_nan'])
    
    
    # Hacemos un loop para que itere sobre las columnas del data frame
    for columna in dataframe:
        
        # Condicional para datos numericos
        if pd.api.types.is_numeric_dtype(dataframe[columna]):
            # Estadisticos
            tipo_de_dato = dataframe[columna].dtype 
            obs = len(dataframe[columna])
            std = dataframe[columna].std()
            media = dataframe[columna].mean()
            cuantil_25 = dataframe[columna].quantile(.25)
            cuantil_50 = dataframe[columna].quantile(.5)
            cuantil_75 = dataframe[columna].quantile(.75)
            maximo = dataframe[columna].max()
            minimo = dataframe[columna].min()
            obs_unic = dataframe[columna].nunique()
            num_de_nan = dataframe[columna].isna().sum()
            # Top 3
            if dataframe[columna].nunique() >= 3:
                top_3= dataframe.groupby([columna])[columna].count().sort_values(ascending = False).head(3)
                top_1 = top_3.index[0]
                top_2 = top_3.index[1]
                top_3 = top_3.index[2]
            else:
                top_3 = dataframe.groupby([columna])[columna].count().sort_values(ascending = False).head(3)
                top_1 = top_3.index[0]
                top_2 = top_3.index[0]
                top_3 = top_3.index[0]
                
            # Asignamos los estadisticos a las columnas
            datos = {'tipo_de_dato':tipo_de_dato,'observaciones':obs,'media':media,
                    'cuantil_25':cuantil_25,'cuantil_50':cuantil_50,'cuantil_75':cuantil_75,
                    'valor_max':maximo,'valor_min':minimo,'observaciones_unicas':obs_unic,
                    'top_1':top_1,'top_2':top_2,'top_3':top_3,'numero_de_nan':num_de_nan}
            
            # Los vamos agregando a un nuevo data frame 
            df1 = pd.DataFrame(datos, index = [columna])
            df_num = df_num.append(df1)
        
        # Condicional para los categoricos    
        else:
            # Estadisticos
            num_cat = dataframe[columna].nunique()
            moda = dataframe[columna].mode(dropna=True)[0]
            num_de_nan = dataframe[columna].isna().sum()
            # Top 3
            if dataframe[columna].nunique() >= 3:
                top_3= dataframe.groupby([columna])[columna].count().sort_values(ascending = False).head(3)
                top_1 = top_3.index[0]
                top_2 = top_3.index[1]
                top_3 = top_3.index[2]
            else:
                top_3 = dataframe.groupby([columna])[columna].count().sort_values(ascending = False).head(3)
                top_1 = top_3.index[0]
                top_2 = top_3.index[0]
                top_3 = top_3.index[0]
                
            # Asignamos estadisticos a columnas
            datos = {'n_de_categorias':num_cat,'moda':moda,'top_1':top_1,'top_2':top_2,
                     'top_3':top_3,'numero_de_nan':num_de_nan}
            
            # Los vamos agregando a un nuevo data frame
            df1 = pd.DataFrame(datos, index = [columna])
            df_cat = df_cat.append(df1)
            
    # Regresa dos bases una para las variables numericas y otro para variables categoricas
    return df_num, df_cat

In [48]:
numerico, categorico = data_profiling(df)

In [49]:
display(numerico)

tipo_de_dato observaciones         media    cuantil_25  \
consumo_total_mixto       float64         71102    174.359929      0.000000   
anio                        int64         71102   2019.000000   2019.000000   
consumo_prom_dom          float64         71102     29.132386     18.690547   
consumo_total_dom         float64         71102   1186.263611    161.635000   
consumo_prom_mixto        float64         71102     50.636234      0.000000   
consumo_total             float64         71102   1695.847222    340.952500   
consumo_prom              float64         71102    111.217399     23.010139   
consumo_prom_no_dom       float64         71102    126.760172      6.275417   
consumo_total_no_dom      float64         71102    436.060309     10.980000   
gid                         int64         71102  35551.500000  17776.250000   

                        cuantil_50    cuantil_75      valor_max  valor_min  \
consumo_total_mixto      79.940000    233.320000   23404.440000        0.0   
anio                   2019.000000   2019.000000    2019.000000     2019.0   
consumo_prom_dom         26.414248     36.246563    7796.410000        0.0   
consumo_total_dom       604.185000   1261.445000   95060.690000        0.0   
consumo_prom_mixto       33.451667     61.216548   11702.220000        0.0   
consumo_total           896.175000   1808.902500  119726.940000        0.0   
consumo_prom             31.693818     45.484917   89691.773438        0.0   
consumo_prom_no_dom      19.280000     54.186875   89691.773438        0.0   
consumo_total_no_dom     54.055000    230.430000  119726.940000        0.0   
gid                   35551.500000  53326.750000   71102.000000        1.0   

                     observaciones_unicas    top_1     top_2     top_3  \
consumo_total_mixto                 24339      0.0     36.00     17.70   
anio                                    1   2019.0   2019.00   2019.00   
consumo_prom_dom                    52060      0.0      1.22     14.64   
consumo_total_dom                   47051      0.0      1.22     10.98   
consumo_prom_mixto                  31911      0.0     36.00     29.28   
consumo_total                       56015      0.0      3.05      1.22   
consumo_prom                        62214      0.0      1.22      3.05   
consumo_prom_no_dom                 37440      0.0      1.22      1.83   
consumo_total_no_dom                27336      0.0      1.22      1.83   
gid                                 71102  71102.0  23703.00  23697.00   

                     numero_de_nan  
consumo_total_mixto           8327  
anio                             0  
consumo_prom_dom              4820  
consumo_total_dom             4820  
consumo_prom_mixto            8327  
consumo_total                    0  
consumo_prom                     0  
consumo_prom_no_dom              0  
consumo_total_no_dom             0  
gid                              0

### Variables categóricas:

- Número de categorías
- Valor de las categorías
- Valores faltantes
- Moda
- Número de observaciones por categoría
- Proporción de observaciones por categoría. 
- Top 1, top 2, top 3 (moda 1, moda 2, moda 3)

In [60]:
# Número de categorías, moda, top 3 y valores faltantes
display(categorico)

n_de_categorias                                               moda  \
Geo Point            22930                        19.300946991,-99.1582172757   
Geo Shape            22922  {"type": "MultiPolygon", "coordinates": [[[[-9...   
nomgeo                  17                                         Iztapalapa   
alcaldia                16                                         IZTAPALAPA   
colonia               1340                                             CENTRO   
bimestre                 3                                                  2   
indice_des               4                                               BAJO   

                                                        top_1  \
Geo Point                        19.5112685166,-99.1375628607   
Geo Shape   {"type": "MultiPolygon", "coordinates": [[[[-9...   
nomgeo                                             Iztapalapa   
alcaldia                                           IZTAPALAPA   
colonia                                                CENTRO   
bimestre                                                    2   
indice_des                                               BAJO   

                                                        top_2  \
Geo Point                        19.5143312067,-99.1436926134   
Geo Shape   {"type": "MultiPolygon", "coordinates": [[[[-9...   
nomgeo                                      Gustavo A. Madero   
alcaldia                                    GUSTAVO A. MADERO   
colonia                                     AGRICOLA ORIENTAL   
bimestre                                                    3   
indice_des                                            POPULAR   

                                                        top_3 numero_de_nan  
Geo Point                        19.5031486487,-99.0890349339             0  
Geo Shape   {"type": "MultiPolygon", "coordinates": [[[[-9...            24  
nomgeo                                             Cuauhtémoc             0  
alcaldia                                           CUAUHTEMOC             0  
colonia                                            ROMA NORTE             0  
bimestre                                                    1             0  
indice_des                                               ALTO             0

In [51]:
# Valor de las categorías
df.colonia.unique()

array(['7 DE NOVIEMBRE', 'NUEVA TENOCHTITLAN', 'PROHOGAR', ...,
       'FERROCARRILERA INSURGENTES', 'AMPLIACION CASTILLO GRANDE',
       'PALO ALTO'], dtype=object)

In [52]:
df.bimestre.unique()

array(['3', '1', '2'], dtype=object)

In [53]:
df.indice_des.unique()

array(['ALTO', 'MEDIO', 'POPULAR', 'BAJO'], dtype=object)

In [54]:
# Número de observaciones por categoría y proporción de observaciones por categoría. 

def num_prop(column_data):
    # Número de observaciones por categoría
    num=pd.value_counts(df[column_data]) 
    # Proporción de observaciones por categoría
    prop=100 * df[column_data].value_counts() / len(df[column_data])
    # Unión de columnas en una y renombramiento
    columnas_num_prop = round(pd.concat([num, prop], 
                                        keys=['num_obs_cat', 'prop_obs_cat'],
                                        axis=1),2)
    
    return columnas_num_prop

In [55]:
num_prop('nomgeo')

num_obs_cat  prop_obs_cat
Iztapalapa                    10515         14.79
Gustavo A. Madero             10058         14.15
Cuauhtémoc                     7313         10.29
Benito Juárez                  6049          8.51
Venustiano Carranza            5179          7.28
Miguel Hidalgo                 5110          7.19
Coyoacán                       4947          6.96
Azcapotzalco                   4216          5.93
Álvaro Obregón                 4140          5.82
Iztacalco                      3469          4.88
Xochimilco                     2450          3.45
Talpan                         2140          3.01
Tláhuac                        1955          2.75
Tlalpan                        1064          1.50
La Magdalena Contreras          955          1.34
Cuajimalpa de Morelos           892          1.25
Milpa Alta                      650          0.91

In [56]:
num_prop('alcaldia')

num_obs_cat  prop_obs_cat
IZTAPALAPA                 10515         14.79
GUSTAVO A. MADERO          10058         14.15
CUAUHTEMOC                  7313         10.29
BENITO JUAREZ               6049          8.51
VENUSTIANO CARRANZA         5179          7.28
MIGUEL HIDALGO              5110          7.19
COYOACAN                    4947          6.96
AZCAPOTZALCO                4216          5.93
ALVARO OBREGON              4140          5.82
IZTACALCO                   3469          4.88
TLALPAN                     3204          4.51
XOCHIMILCO                  2450          3.45
TLAHUAC                     1955          2.75
MAGDALENA CONTRERAS          955          1.34
CUAJIMALPA                   892          1.25
MILPA ALTA                   650          0.91

In [57]:
num_prop('colonia')

num_obs_cat  prop_obs_cat
CENTRO                       1139          1.60
AGRICOLA ORIENTAL             837          1.18
ROMA NORTE                    602          0.85
MOCTEZUMA 2A SECCION          558          0.78
JARDIN BALBUENA               498          0.70
...                           ...           ...
HUIZACHITO                      2          0.00
CALZADA JALALPA                 2          0.00
PIRU SECC. I                    2          0.00
SANTISIMA TRINIDAD              1          0.00
U. HAB. NUEVA ROSITA            1          0.00

[1340 rows x 2 columns]

In [58]:
num_prop('bimestre')

num_obs_cat  prop_obs_cat
2        23942         33.67
3        23822         33.50
1        23338         32.82

In [59]:
num_prop('indice_des')

num_obs_cat  prop_obs_cat
BAJO           29248         41.14
POPULAR        16539         23.26
ALTO           15516         21.82
MEDIO           9799         13.78

#### 8. ¿Qué conocemos ahora de este set de datos por variable?

Sabemos que tenemos disponibles 17 variables, y 71102 observaciones. De las cuales 7 son variables continuas, 5 son categóricas. En el caso de geo_point la tenemos como string, pero será necesario separarla en latitud y longitud para después convertirla a una variable numérica. Adicionalmente, se observa que nomgeo y alcaldía son variables que contienen la misma información.

Tenemos información de sólo el 2019 y de 1340 colonias donde la colonia Centro tiene más observaciones, seguida de la Agricola Oriental y la Roma Norte. Por otro lado, las delegaciones con más observaciones son Iztapalapa y Gustavo A. Madero. En la variable indice_des categoría más común es la de Bajo, con alrededor de un 40% de observaciones. Los tres tipos de bimestres tienen proporciones muy similares. 

#### 10. ¿Cuántas alcadías tienes? ¿Cuántos nomgeo tienes? ¿Identificas algún error?

In [39]:
df.alcaldia.unique()

array(['GUSTAVO A. MADERO', 'AZCAPOTZALCO', 'COYOACAN', 'TLALPAN',
       'MILPA ALTA', 'ALVARO OBREGON', 'CUAJIMALPA', 'TLAHUAC',
       'MAGDALENA CONTRERAS', 'XOCHIMILCO', 'IZTAPALAPA', 'IZTACALCO',
       'BENITO JUAREZ', 'MIGUEL HIDALGO', 'CUAUHTEMOC',
       'VENUSTIANO CARRANZA'], dtype=object)

In [40]:
df.alcaldia.nunique()

16

In [41]:
df.nomgeo.unique()

array(['Gustavo A. Madero', 'Azcapotzalco', 'Coyoacán', 'Talpan',
       'Milpa Alta', 'Álvaro Obregón', 'Cuajimalpa de Morelos', 'Tláhuac',
       'La Magdalena Contreras', 'Xochimilco', 'Iztapalapa', 'Iztacalco',
       'Benito Juárez', 'Miguel Hidalgo', 'Cuauhtémoc',
       'Venustiano Carranza', 'Tlalpan'], dtype=object)

In [42]:
df.nomgeo.nunique()

17

 Si vemos los agrupados con sus conteos notamos que ambas variables contienen el nombre de las delegaciones de la CDMX; sin embargo, una de las variables tiene los nombres de estas en mayúsculas y la otra las tiene en minúsculas, y además una de ellas tiene acentos y la otra no. Adicionalmente notemos que nomgeo tiene 17, mientras que alcaldias tiene 16, ya que la primera trae mal escrito el registro "Talpan" (sí existe un lugar llamado así, pero no es una delegación de la CDMX, por lo que podemos asumir que este dato fue inputado de manera errónea).
 Si agrupamos por alcaldia y nomgeo es evidente que la suma de "Talpan" y "Tlalpan" (en nomgeo) contienen la misma cantidad de registros que "TLALPAN" (en alcaldias), por lo que podemos concluir que sí es un error:

In [43]:
df.groupby(['alcaldia']).count()

Geo Point  Geo Shape  consumo_total_mixto   anio  nomgeo  \
alcaldia                                                                        
ALVARO OBREGON            4140       4137                 3520   4140    4140   
AZCAPOTZALCO              4216       4216                 3921   4216    4216   
BENITO JUAREZ             6049       6046                 5594   6049    6049   
COYOACAN                  4947       4941                 4065   4947    4947   
CUAJIMALPA                 892        892                  746    892     892   
CUAUHTEMOC                7313       7310                 6726   7313    7313   
GUSTAVO A. MADERO        10058      10058                 9187  10058   10058   
IZTACALCO                 3469       3469                 3278   3469    3469   
IZTAPALAPA               10515      10512                 8719  10515   10515   
MAGDALENA CONTRERAS        955        955                  757    955     955   
MIGUEL HIDALGO            5110       5104                 4472   5110    5110   
MILPA ALTA                 650        650                  562    650     650   
TLAHUAC                   1955       1955                 1692   1955    1955   
TLALPAN                   3204       3204                 2492   3204    3204   
VENUSTIANO CARRANZA       5179       5179                 4831   5179    5179   
XOCHIMILCO                2450       2450                 2213   2450    2450   

                     consumo_prom_dom  consumo_total_dom  colonia  \
alcaldia                                                            
ALVARO OBREGON                   3944               3944     4140   
AZCAPOTZALCO                     4044               4044     4216   
BENITO JUAREZ                    5940               5940     6049   
COYOACAN                         4517               4517     4947   
CUAJIMALPA                        835                835      892   
CUAUHTEMOC                       6962               6962     7313   
GUSTAVO A. MADERO                9520               9520    10058   
IZTACALCO                        3405               3405     3469   
IZTAPALAPA                       8937               8937    10515   
MAGDALENA CONTRERAS               830                830      955   
MIGUEL HIDALGO                   4933               4933     5110   
MILPA ALTA                        562                562      650   
TLAHUAC                          1764               1764     1955   
TLALPAN                          2717               2717     3204   
VENUSTIANO CARRANZA              5026               5026     5179   
XOCHIMILCO                       2346               2346     2450   

                     consumo_prom_mixto  consumo_total  consumo_prom  \
alcaldia                                                               
ALVARO OBREGON                     3520           4140          4140   
AZCAPOTZALCO                       3921           4216          4216   
BENITO JUAREZ                      5594           6049          6049   
COYOACAN                           4065           4947          4947   
CUAJIMALPA                          746            892           892   
CUAUHTEMOC                         6726           7313          7313   
GUSTAVO A. MADERO                  9187          10058         10058   
IZTACALCO                          3278           3469          3469   
IZTAPALAPA                         8719          10515         10515   
MAGDALENA CONTRERAS                 757            955           955   
MIGUEL HIDALGO                     4472           5110          5110   
MILPA ALTA                          562            650           650   
TLAHUAC                            1692           1955          1955   
TLALPAN                            2492           3204          3204   
VENUSTIANO CARRANZA                4831           5179          5179   
XOCHIMILCO                         2213           2450          2450   

                     c

In [44]:
df.groupby(['alcaldia', 'nomgeo']).count()

Geo Point  Geo Shape  \
alcaldia            nomgeo                                         
ALVARO OBREGON      Álvaro Obregón               4140       4137   
AZCAPOTZALCO        Azcapotzalco                 4216       4216   
BENITO JUAREZ       Benito Juárez                6049       6046   
COYOACAN            Coyoacán                     4947       4941   
CUAJIMALPA          Cuajimalpa de Morelos         892        892   
CUAUHTEMOC          Cuauhtémoc                   7313       7310   
GUSTAVO A. MADERO   Gustavo A. Madero           10058      10058   
IZTACALCO           Iztacalco                    3469       3469   
IZTAPALAPA          Iztapalapa                  10515      10512   
MAGDALENA CONTRERAS La Magdalena Contreras        955        955   
MIGUEL HIDALGO      Miguel Hidalgo               5110       5104   
MILPA ALTA          Milpa Alta                    650        650   
TLAHUAC             Tláhuac                      1955       1955   
TLALPAN             Talpan                       2140       2140   
                    Tlalpan                      1064       1064   
VENUSTIANO CARRANZA Venustiano Carranza          5179       5179   
XOCHIMILCO          Xochimilco                   2450       2450   

                                            consumo_total_mixto   anio  \
alcaldia            nomgeo                                               
ALVARO OBREGON      Álvaro Obregón                         3520   4140   
AZCAPOTZALCO        Azcapotzalco                           3921   4216   
BENITO JUAREZ       Benito Juárez                          5594   6049   
COYOACAN            Coyoacán                               4065   4947   
CUAJIMALPA          Cuajimalpa de Morelos                   746    892   
CUAUHTEMOC          Cuauhtémoc                             6726   7313   
GUSTAVO A. MADERO   Gustavo A. Madero                      9187  10058   
IZTACALCO           Iztacalco                              3278   3469   
IZTAPALAPA          Iztapalapa                             8719  10515   
MAGDALENA CONTRERAS La Magdalena Contreras                  757    955   
MIGUEL HIDALGO      Miguel Hidalgo                         4472   5110   
MILPA ALTA          Milpa Alta                              562    650   
TLAHUAC             Tláhuac                                1692   1955   
TLALPAN             Talpan                                 2140   2140   
                    Tlalpan                                 352   1064   
VENUSTIANO CARRANZA Venustiano Carranza                    4831   5179   
XOCHIMILCO          Xochimilco                             2213   2450   

                                            consumo_prom_dom  \
alcaldia            nomgeo                                     
ALVARO OBREGON      Álvaro Obregón                      3944   
AZCAPOTZALCO        Azcapotzalco                        4044   
BENITO JUAREZ       Benito Juárez                       5940   
COYOACAN            Coyoacán                            4517   
CUAJIMALPA          Cuajimalpa de Morelos                835   
CUAUHTEMOC          Cuauhtémoc                          6962   
GUSTAVO A. MADERO   Gustavo A. Madero                   9520   
IZTACALCO           Iztacalco                           3405   
IZTAPALAPA          Iztapalapa                          8937   
MAGDALENA CONTRERAS La Magdalena Contreras               830   
MIGUEL HIDALGO      Miguel Hidalgo                      4933   
MILPA ALTA          Milpa Alta                           562   
TLAHUAC             Tláhuac                             1764   
TLALPAN             Talpan                              2140   
                    Tlalpan                              577   
VENUSTIANO CARRANZA Venustiano Carranza                 5026   
XOCHIMILCO          Xochimilco                          2346   

                                            consumo_total_dom  colonia  \
alcaldia            nomgeo                           

#### 11. Transformar el nombre de las columnas a formato estándar: minúsculas, sin espacios en blanco -cambiar por guónes bajos-, sin signos de puntuación.

In [45]:
list(df.columns)

['Geo Point',
 'Geo Shape',
 'consumo_total_mixto',
 'anio',
 'nomgeo',
 'consumo_prom_dom',
 'consumo_total_dom',
 'alcaldia',
 'colonia',
 'consumo_prom_mixto',
 'consumo_total',
 'consumo_prom',
 'consumo_prom_no_dom',
 'bimestre',
 'consumo_total_no_dom',
 'gid',
 'indice_des']

In [46]:
df = df.rename(columns={'Geo Point':'geo_point', 'Geo Shape':'geo_shape'})

In [47]:
list(df.columns)

['geo_point',
 'geo_shape',
 'consumo_total_mixto',
 'anio',
 'nomgeo',
 'consumo_prom_dom',
 'consumo_total_dom',
 'alcaldia',
 'colonia',
 'consumo_prom_mixto',
 'consumo_total',
 'consumo_prom',
 'consumo_prom_no_dom',
 'bimestre',
 'consumo_total_no_dom',
 'gid',
 'indice_des']

#### 12. Agregar la varaible latitud y longitud

In [48]:
df.geo_point

0        19.4552601937,-99.1126617526
1        19.4552601937,-99.1126617526
2        19.4557195871,-99.1135822797
3        19.4596467168,-99.1044693641
4        19.4741606185,-99.1467497317
                     ...             
71097    19.4485642979,-99.1399395353
71098    19.4493393649,-99.1457191092
71099    19.4483923147,-99.1459300721
71100    19.4475868325,-99.1425094385
71101    19.4474017534,-99.1397251034
Name: geo_point, Length: 71102, dtype: object

In [49]:
nuevo = df.geo_point.str.split(",", n=1, expand = True)

In [50]:
nuevo

0               1
0      19.4552601937  -99.1126617526
1      19.4552601937  -99.1126617526
2      19.4557195871  -99.1135822797
3      19.4596467168  -99.1044693641
4      19.4741606185  -99.1467497317
...              ...             ...
71097  19.4485642979  -99.1399395353
71098  19.4493393649  -99.1457191092
71099  19.4483923147  -99.1459300721
71100  19.4475868325  -99.1425094385
71101  19.4474017534  -99.1397251034

[71102 rows x 2 columns]

In [53]:
df['latitud'] = nuevo[0]
df['longitud'] = nuevo[1]  # las pone hasta el último

In [54]:
df.head()

geo_point  \
0  19.4552601937,-99.1126617526   
1  19.4552601937,-99.1126617526   
2  19.4557195871,-99.1135822797   
3  19.4596467168,-99.1044693641   
4  19.4741606185,-99.1467497317   

                                           geo_shape  consumo_total_mixto  \
0  {"type": "MultiPolygon", "coordinates": [[[[-9...               159.72   
1  {"type": "MultiPolygon", "coordinates": [[[[-9...                 0.00   
2  {"type": "MultiPolygon", "coordinates": [[[[-9...                 0.00   
3  {"type": "MultiPolygon", "coordinates": [[[[-9...                 0.00   
4  {"type": "MultiPolygon", "coordinates": [[[[-9...                56.72   

   anio             nomgeo  consumo_prom_dom  consumo_total_dom  \
0  2019  Gustavo A. Madero         42.566364             468.23   
1  2019  Gustavo A. Madero         35.936667             107.81   
2  2019  Gustavo A. Madero         24.586000             122.93   
3  2019  Gustavo A. Madero          0.000000               0.00   
4  2019       Azcapotzalco         67.436250             539.49   

            alcaldia             colonia  consumo_prom_mixto  consumo_total  \
0  GUSTAVO A. MADERO      7 DE NOVIEMBRE               53.24         631.00   
1  GUSTAVO A. MADERO      7 DE NOVIEMBRE                0.00         115.13   
2  GUSTAVO A. MADERO      7 DE NOVIEMBRE                0.00         197.96   
3  GUSTAVO A. MADERO  NUEVA TENOCHTITLAN                0.00         253.53   
4       AZCAPOTZALCO            PROHOGAR               56.72         839.35   

   consumo_prom  consumo_prom_no_dom bimestre  consumo_total_no_dom    gid  \
0     42.066667                 3.05        3                  3.05  57250   
1     28.782500                 7.32        3                  7.32  57253   
2     32.993333                75.03        3                 75.03  57255   
3     84.510000                84.51        3                253.53  57267   
4     76.304545               121.57        3                243.14  57330   

  indice_des        latitud        longitud  
0       ALTO  19.4552601937  -99.1126617526  
1      MEDIO  19.4552601937  -99.1126617526  
2    POPULAR  19.4557195871  -99.1135822797  
3       BAJO  19.4596467168  -99.1044693641  
4       BAJO  19.4741606185  -99.1467497317

#### 13. Pasar la variable latitud y longitud a numérica -si no la tomó como numérica-.

Notemos que las variables se quedaron como tipo string:

In [55]:
print(type(df['longitud'][1]))
print(type(df['latitud'][1]))

<class 'str'>
<class 'str'>


Las transromamos a variables numéricas: 

In [56]:
df['latitud'] = df.latitud.astype('float64')
df['longitud'] = df.longitud.astype('float64')

Y verificamos que se haya cambiado el tipo de dato:

In [57]:
print(type(df['longitud'][1]))
print(type(df['latitud'][1]))

<class 'numpy.float64'>
<class 'numpy.float64'>


#### 14. Eliminar la columna geo_point -una vez que create la variable latitud y longitud.

Eliminamos la columna geo_point con un drop:

In [58]:
df.drop(columns = ['geo_point'], inplace = True)

In [59]:
df.head()

geo_shape  consumo_total_mixto  \
0  {"type": "MultiPolygon", "coordinates": [[[[-9...               159.72   
1  {"type": "MultiPolygon", "coordinates": [[[[-9...                 0.00   
2  {"type": "MultiPolygon", "coordinates": [[[[-9...                 0.00   
3  {"type": "MultiPolygon", "coordinates": [[[[-9...                 0.00   
4  {"type": "MultiPolygon", "coordinates": [[[[-9...                56.72   

   anio             nomgeo  consumo_prom_dom  consumo_total_dom  \
0  2019  Gustavo A. Madero         42.566364             468.23   
1  2019  Gustavo A. Madero         35.936667             107.81   
2  2019  Gustavo A. Madero         24.586000             122.93   
3  2019  Gustavo A. Madero          0.000000               0.00   
4  2019       Azcapotzalco         67.436250             539.49   

            alcaldia             colonia  consumo_prom_mixto  consumo_total  \
0  GUSTAVO A. MADERO      7 DE NOVIEMBRE               53.24         631.00   
1  GUSTAVO A. MADERO      7 DE NOVIEMBRE                0.00         115.13   
2  GUSTAVO A. MADERO      7 DE NOVIEMBRE                0.00         197.96   
3  GUSTAVO A. MADERO  NUEVA TENOCHTITLAN                0.00         253.53   
4       AZCAPOTZALCO            PROHOGAR               56.72         839.35   

   consumo_prom  consumo_prom_no_dom bimestre  consumo_total_no_dom    gid  \
0     42.066667                 3.05        3                  3.05  57250   
1     28.782500                 7.32        3                  7.32  57253   
2     32.993333                75.03        3                 75.03  57255   
3     84.510000                84.51        3                253.53  57267   
4     76.304545               121.57        3                243.14  57330   

  indice_des    latitud   longitud  
0       ALTO  19.455260 -99.112662  
1      MEDIO  19.455260 -99.112662  
2    POPULAR  19.455720 -99.113582  
3       BAJO  19.459647 -99.104469  
4       BAJO  19.474161 -99.146750

#### 15. Eliminar la columna geo_shape.

In [60]:
df.drop(columns = ['geo_shape'], inplace = True)

In [61]:
df.head()

consumo_total_mixto  anio             nomgeo  consumo_prom_dom  \
0               159.72  2019  Gustavo A. Madero         42.566364   
1                 0.00  2019  Gustavo A. Madero         35.936667   
2                 0.00  2019  Gustavo A. Madero         24.586000   
3                 0.00  2019  Gustavo A. Madero          0.000000   
4                56.72  2019       Azcapotzalco         67.436250   

   consumo_total_dom           alcaldia             colonia  \
0             468.23  GUSTAVO A. MADERO      7 DE NOVIEMBRE   
1             107.81  GUSTAVO A. MADERO      7 DE NOVIEMBRE   
2             122.93  GUSTAVO A. MADERO      7 DE NOVIEMBRE   
3               0.00  GUSTAVO A. MADERO  NUEVA TENOCHTITLAN   
4             539.49       AZCAPOTZALCO            PROHOGAR   

   consumo_prom_mixto  consumo_total  consumo_prom  consumo_prom_no_dom  \
0               53.24         631.00     42.066667                 3.05   
1                0.00         115.13     28.782500                 7.32   
2                0.00         197.96     32.993333                75.03   
3                0.00         253.53     84.510000                84.51   
4               56.72         839.35     76.304545               121.57   

  bimestre  consumo_total_no_dom    gid indice_des    latitud   longitud  
0        3                  3.05  57250       ALTO  19.455260 -99.112662  
1        3                  7.32  57253      MEDIO  19.455260 -99.112662  
2        3                 75.03  57255    POPULAR  19.455720 -99.113582  
3        3                253.53  57267       BAJO  19.459647 -99.104469  
4        3                243.14  57330       BAJO  19.474161 -99.146750

### 16. Cambiar las minúsculas las columnas de alcaldía, colonia e incide_des.

Verifiquemos que las variables sean de tipo string, y después las transformamos a minúsculas:

In [62]:
print(type(df['alcaldia'][1]))
print(type(df['colonia'][1]))
print(type(df['indice_des'][1]))

<class 'str'>
<class 'str'>
<class 'str'>


In [63]:
df['alcaldia'] = df['alcaldia'].str.lower()
df['colonia'] = df['colonia'].str.lower()
df['indice_des'] = df['indice_des'].str.lower()

In [64]:
df.head()

consumo_total_mixto  anio             nomgeo  consumo_prom_dom  \
0               159.72  2019  Gustavo A. Madero         42.566364   
1                 0.00  2019  Gustavo A. Madero         35.936667   
2                 0.00  2019  Gustavo A. Madero         24.586000   
3                 0.00  2019  Gustavo A. Madero          0.000000   
4                56.72  2019       Azcapotzalco         67.436250   

   consumo_total_dom           alcaldia             colonia  \
0             468.23  gustavo a. madero      7 de noviembre   
1             107.81  gustavo a. madero      7 de noviembre   
2             122.93  gustavo a. madero      7 de noviembre   
3               0.00  gustavo a. madero  nueva tenochtitlan   
4             539.49       azcapotzalco            prohogar   

   consumo_prom_mixto  consumo_total  consumo_prom  consumo_prom_no_dom  \
0               53.24         631.00     42.066667                 3.05   
1                0.00         115.13     28.782500                 7.32   
2                0.00         197.96     32.993333                75.03   
3                0.00         253.53     84.510000                84.51   
4               56.72         839.35     76.304545               121.57   

  bimestre  consumo_total_no_dom    gid indice_des    latitud   longitud  
0        3                  3.05  57250       alto  19.455260 -99.112662  
1        3                  7.32  57253      medio  19.455260 -99.112662  
2        3                 75.03  57255    popular  19.455720 -99.113582  
3        3                253.53  57267       bajo  19.459647 -99.104469  
4        3                243.14  57330       bajo  19.474161 -99.146750

#### 17. Volver a correr el proceso de identificación de variables numéricas, categóricas, texto y fechas.

In [65]:
len(list(df.columns))

17

De las cuales, ahora tenemos 12 variables numéricas y 5 de tipo objeto (string):

In [66]:
newdf = df.select_dtypes(include=numerics)
newdf.shape[1]

12

In [67]:
newdf_str = df.select_dtypes(include=['object'])
newdf_str.shape[1]

5

In [68]:
profile = df.profile_report(title = 'Data Profiling con Nuevas Variables')

In [69]:
profile.to_file(output_file = 'pandas_profiling_nuevas_variables.html')